enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedSinglePrimitivePipe.scala

```diff
@@ -0,0 +1,74 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime.slotted.pipes
+
+import org.neo4j.collection.primitive.Primitive
+import org.neo4j.cypher.internal.compatibility.v3_4.runtime.{Slot, SlotConfiguration}
+import org.neo4j.cypher.internal.runtime.PrefetchingIterator
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+import org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.Expression
+import org.neo4j.cypher.internal.runtime.interpreted.pipes.{Pipe, PipeWithSource, QueryState}
+import org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext
+import org.neo4j.cypher.internal.runtime.slotted.helpers.SlottedPipeBuilderUtils.makeSetValueInSlotFunctionFor
+import org.neo4j.cypher.internal.util.v3_4.attribution.Id
+import org.neo4j.values.AnyValue
+
+case class DistinctSlottedSinglePrimitivePipe(source: Pipe,
+                                              slots: SlotConfiguration,
+                                              toSlot: Slot,
+                                              offset: Int,
+                                              expression: Expression)
+                                             (val id: Id = Id.INVALID_ID)
+  extends PipeWithSource(source) {
+
+  //===========================================================================
+  // Compile-time initializations
+  //===========================================================================
+  private val setInSlot: (ExecutionContext, AnyValue) => Unit = makeSetValueInSlotFunctionFor(toSlot)
+
+  expression.registerOwningPipe(this)
+
+  //===========================================================================
+  // Runtime code
+  //===========================================================================
+  protected def internalCreateResults(input: Iterator[ExecutionContext],
+                                      state: QueryState): Iterator[ExecutionContext] = {
+
+    new PrefetchingIterator[ExecutionContext] {
+      private val seen = Primitive.longSet()
+
+      override def produceNext(): Option[ExecutionContext] = {
+        while (input.nonEmpty) { // Let's pull data until we find something not already seen
+          val next = input.next()
+          val id = next.getLongAt(offset)
+          if (seen.add(id)) {
+            // Found something! Set it as the next element to yield, and exit
+            val outgoing = SlottedExecutionContext(slots)
+            val outputValue = expression(next, state)
+            setInSlot(outgoing, outputValue)
+            return Some(outgoing)
+          }
+        }
```

Comment:
If the runtime cost is negligible it would be nice if these pipes could share an abstract base class to avoid duplicating the control flow code.